cpp/src/arrow/record_batch_test.cc

```diff
@@ -332,4 +332,72 @@ TEST_F(TestRecordBatch, MakeEmpty) {
   ASSERT_EQ(empty->num_rows(), 0);
 }
 
+class TestRecordBatchReader : public TestBase {};
+
+TEST_F(TestRecordBatchReader, RangeForLoop) {
+  const int length = 10;
+
+  auto f0 = field("f0", int32());
+  auto f1 = field("f1", uint8());
+  auto f2 = field("f2", int16());
+
+  auto metadata = key_value_metadata({"foo"}, {"bar"});
+
+  std::vector<std::shared_ptr<Field>> fields = {f0, f1, f2};
+  auto schema = ::arrow::schema({f0, f1, f2});
+  auto schema2 = ::arrow::schema({f0, f1});
+  auto schema3 = ::arrow::schema({f0, f1, f2}, metadata);
+
+  auto a0 = MakeRandomArray<Int32Array>(length);
+  auto a1 = MakeRandomArray<UInt8Array>(length);
+  auto a2 = MakeRandomArray<Int16Array>(length);
+
+  auto b1 = RecordBatch::Make(schema, length, {a0, a1, a2});
+  auto b2 = RecordBatch::Make(schema3, length, {a0, a1, a2});
+  auto b3 = RecordBatch::Make(schema2, length, {a0, a1});
+  auto b4 = RecordBatch::Make(schema, length, {a0, a1, a1});
+
+  std::vector<std::shared_ptr<RecordBatch>> batches = {b1, b2, b3, b4};
+
+  ASSERT_OK_AND_ASSIGN(auto reader, RecordBatchReader::Make(batches));
+
+  int64_t i = 0;
+  for (auto maybe_batch : *reader) {
+    ASSERT_OK_AND_ASSIGN(auto batch, maybe_batch);
+    AssertBatchesEqual(*batch, *batches[i++]);
+  }
+}
+
+TEST_F(TestRecordBatchReader, BeginEndForLoop) {
+  const int length = 10;
+
+  auto field1 = field("f1", int32());
+  auto field2 = field("f2", uint8());
+  auto field3 = field("f3", int16());
```

Comment:
Can you factor out (in the fixture class above) and reuse the test data for these two tests?